File: DataStructuresAndAlgorithms/1. Array and Hashing/2. Valid Anagram/validAnagram.py
from collections import Counter

def isAnagram(s: str, t: str) -> bool:
    if len(s) != len(t):
        return False
    counts = Counter(s)
    for i in t:
        if i in counts:
            if counts[i] <= 0:
                return False
            counts[i] -= 1
        else: return False
    return True

if __name__ == '__main__':
    #s = "rat" t = "car"
    s = "anagram" 
    t = "nagaram"
    print(isAnagram(s, t))
        

File: DataStructuresAndAlgorithms/1. Array and Hashing/2. Valid Anagram/validAnagram.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

bool isAnagram(string s, string t){
    if (s.length() != t.length()) return false;
    unordered_map<char, int> counts;
    for(int i = 0; i < s.length(); i++){
        counts[s[i]]++;
    }
    for(int i = 0; i < t.length(); i++){
        if(counts[t[i]] != 0){
            counts[t[i]]--;
        }
        else return false;
    }
    return true;
}

int main(int argc, char **argv){
    string s = "anagram";
    string t = "nagaram";
    cout << isAnagram(s, t) << endl;
    return 0;
}